plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs

```diff
@@ -0,0 +1,211 @@
+                                                                 (renameType x z a')
+                                                                 (renameType y z b')
+          where z = freshenTyName (locals <> Map.keysSet ctx) x
+        _                                              -> mzero
+      where
+        -- Check that in the current context we can solve variable z to type ty.
+        validSolve z ty = and [ Set.member z flex -- z must be a "meta variable"
+                              -- z can't be a locally bound variable
+                              , not $ Set.member z locals
+                              -- we can't solve z by changing it to a type
+                              -- where `z` is free in the type.
+                              , not $ Set.member z fvs
+                              -- The solve has to be well scoped
+                              , checkKind ctx ty (ctx Map.! z)
+                              -- We can't capture a locally bound variable
+                              , null $ Set.intersection fvs locals ]
+                              where fvs = fvTypeR sub ty
+
+    unifies sub _ [] [] = pure sub
+    unifies sub locals (a : as) (b : bs) = do
+      sub1 <- go sub locals a b
+      unifies sub1 locals as bs
+    unifies _ _ _ _ = mzero
+
+-- | Parallel substitution
+parSubstType :: Map TyName (Type TyName DefaultUni ())
+             -> Type TyName DefaultUni ()
+             -> Type TyName DefaultUni ()
+parSubstType = substType' False
+
+-- CODE REVIEW: this function is a bit strange and I don't like it. Ideas welcome for how to
+-- do this better. It basically deals with the fact that we want to be careful when substituting
+-- the datatypes that escape from a term into the type. It's yucky but it works.
+--
+-- This might not be a welcome opinion, but working with this stuff exposes some of
+-- the shortcomings of the current PIR design. It would be cleaner if a PIR program was a list
+-- of declarations and datatype declarations weren't in terms.
+-- TODO: Is this actually doing anything other than what we already do in other substitution functions?!
+substEscape :: Set TyName
+            -> Map TyName (Type TyName DefaultUni ())
+            -> Type TyName DefaultUni ()
+            -> Type TyName DefaultUni ()
+substEscape fv sub ty = case ty of
+  TyVar _ x        -> maybe ty (substEscape fv sub) (Map.lookup x sub)
+  TyFun _ a b      -> TyFun () (substEscape fv sub a) (substEscape fv sub b)
+  TyApp _ a b      -> TyApp () (substEscape fv sub a) (substEscape fv sub b)
+  TyLam _ x k b    -> TyLam () x k $ substEscape (Set.insert x fv) sub b
+  TyForall _ x k b -> TyForall () x k $ substEscape (Set.insert x fv) sub b
+  TyBuiltin{}      -> ty
+  TyIFix{}         -> ty
+
+-- CODE REVIEW: does this exist anywhere?
+renameType :: TyName -> TyName -> Type TyName DefaultUni () -> Type TyName DefaultUni ()
```

Comment:
Not that I'm aware of. We've managed to avoid doing name-capture-aware substitution.

Reply:
I decided to keep all the substitution stuff in the generators modules. We'll pull things out if we ever need that.